integrations/tensorflow/bindings/python/pyiree/tf/support/tf_test_utils.py

```diff
@@ -386,3 +409,205 @@ def setUp(self):
+
+  def setUp(self):
+    # Ran before each unit test.
+    super().setUp()
+    self._ref_module.create_reinitialized()
+    self._tar_modules = [
+        module.create_reinitialized() for module in self._tar_modules
+    ]
+
+  def compare_backends(self, trace_function, rtol=1e-6, atol=1e-6):
+    """Run the reference and target backends on trace_function and compare them.
+
+    Args:
+      trace_function: a function accepting a TracedModule as its argument.
+      rtol: a float as defined in np.allclose. Will be overridden in any
+        comparison between calls to the trace with a `rtol` kwarg.
+      atol: a float as defined in np.allclose. Will be overridden in any
+        comparison between calls to the trace with a `atol` kwarg.
+    """
+    # Trace the test function for each backend.
+    ref_trace = tf_utils.TracedModule(self._ref_module, trace_function)
+    tar_traces = [
+        tf_utils.TracedModule(module, trace_function)
+        for module in self._tar_modules
+    ]
+
+    # Compare each target trace with the reference trace.
+    failed_backend_indices = []
+    for i, tar_trace in enumerate(tar_traces):
+      logging.info("Comparing the reference backend '%s' with '%s'",
+                   ref_trace.backend, tar_trace.backend)
+      traces_match = self._compare_traces(ref_trace, tar_trace, rtol, atol)
+      if not traces_match:
+        failed_backend_indices.append(i)
+
+    # Save the results to disk before validating.
+    trace_dir = os.path.join(self._artifacts_dir, "traces")
+    if not os.path.exists(trace_dir):
+      os.makedirs(trace_dir)
+    ref_trace.save_plaintext(trace_dir, FLAGS.summarize)
+    for tar_trace in tar_traces:
+      tar_trace.save_plaintext(trace_dir, FLAGS.summarize)
+
+    # Validate results.
+    if len(failed_backend_indices) > 0:
+      # Extract info for logging.
+      failed_traces = [tar_traces[b] for b in failed_backend_indices]
+      failed_backends = [trace.backend for trace in failed_traces]
+      failure_info = (
+          "Comparision between the reference backend and the following targets "
+          f"failed: {failed_backends}. The  errors above show the outputs of "
+          "the non-matching calls.")
+
+      # This condition is always True, but is useful for context in the logs.
+      self.assertFalse(len(failed_backends) > 0, failure_info)
+
+  def _compare_traces(self, ref_trace, tar_trace, default_rtol, default_atol):
+    traces_match = True
+    for ref_call, tar_call in zip(ref_trace, tar_trace):
+      ref_call.record_tolerances_if_unset(default_rtol, default_atol)
+      tar_call.record_tolerances_if_unset(default_rtol, default_atol)
+      # Get the potentially overridden tolerances for comparisons of this call.
+      rtol, atol = ref_call.get_tolerances()
+
+      logging.info("Comparing calls to '%s'", ref_call.method)
+      calls_match = self._check_same(ref_call.outputs, tar_call.outputs, rtol,
```

Comment:
we should also check the inputs, just to be safe. Certain kinds of "misuse" of the testing framework can result in traces that don't make sense to compare. e.g.

m.foo(np.array(1.0) if rand() > 0.5 else np.array(2.0))

Reply:
Sounds good. The same result for different inputs could be misleading in some circumstances too.